pkg/image/external/external.go

```diff
@@ -0,0 +1,140 @@
+	if source == "" {
+		source = UNKNOWN
+	}
+
+	logrus.Infof("generating %s image list...", source)
+	externalImagesMap := make(map[string]bool)
+	releases, _ := externalData["releases"].([]interface{})
+	var compatibleReleases []string
+
+	for _, release := range releases {
+		releaseMap, _ := release.(map[string]interface{})
+		version, _ := releaseMap["version"].(string)
+		if version == "" {
+			continue
+		}
+
+		if rancherVersion != "dev" {
+			maxVersion, _ := releaseMap["maxChannelServerVersion"].(string)
+			maxVersion = strings.TrimPrefix(maxVersion, "v")
+			if maxVersion == "" {
+				continue
+			}
+			minVersion, _ := releaseMap["minChannelServerVersion"].(string)
+			minVersion = strings.Trim(minVersion, "v")
+			if minVersion == "" {
+				continue
+			}
+
+			// IsNewerVersion works for both k3s and rke2.
+			versionGTMin, err := k3sbasedupgrade.IsNewerVersion(minVersion, rancherVersion)
+			if err != nil {
+				continue
+			}
+			if rancherVersion != minVersion && !versionGTMin {
+				// Rancher version not equal to or greater than minimum supported rancher version.
+				continue
+			}
+
+			// IsNewerVersion works for both k3s and rke2.
+			versionLTMax, err := k3sbasedupgrade.IsNewerVersion(rancherVersion, maxVersion)
+			if err != nil {
+				continue
+			}
+			if rancherVersion != maxVersion && !versionLTMax {
+				// Rancher version not equal to or greater than maximum supported rancher version.
+				continue
+			}
+		}
+
+		compatibleReleases = append(compatibleReleases, version)
+	}
+
+	for _, release := range compatibleReleases {
+		// Registries don't allow "+", so image names will have these substituted.
+		upgradeImage := fmt.Sprintf("rancher/%s-upgrade:%s", source, strings.Replace(release, "+", "-", -1))
+		externalImagesMap[upgradeImage] = true
+
+		images, err := downloadExternalSupportingImages(release, source, os, arch)
+		if err != nil {
+			logrus.Infof("could not find supporting images for %s release [%s]: %v", source, release, err)
+			continue
+		}
+
+		supportingImages := strings.Split(images, "\n")
+		if supportingImages[len(supportingImages)-1] == "" {
+			supportingImages = supportingImages[:len(supportingImages)-1]
+		}
+
+		for _, imageName := range supportingImages {
+			imageName = strings.TrimPrefix(imageName, "docker.io/")
+			externalImagesMap[imageName] = true
+		}
+	}
+
+	var externalImages []string
+	for imageName := range externalImagesMap {
+		externalImages = append(externalImages, imageName)
+	}
+
+	sort.Strings(externalImages)
+	logrus.Infof("finished generating %s image list...", source)
+	return externalImages
+}
+
+func downloadExternalSupportingImages(release string, source Source, os string, arch string) (string, error) {
+	var url string
+	switch source {
+	case RKE2:
+		if os == "" {
+			os = "linux"
+		}
+		if arch == "" {
+			arch = "amd64"
+		}
+		url = fmt.Sprintf("https://github.com/rancher/rke2/releases/download/%s/rke2-images.%s-%s.txt", release, os, arch)
```

Comment:
Using `rke2-images` or `rke2-images-all` is subject to change based on internal discussion. Other than that, this url will remain the same.

Reply:
**Update**: this needs to be `rke2-images-all`, but that is only available in RKE2 with Kubernetes v1.21+.